quinn-proto/src/range_set/array_range_set.rs

```diff
@@ -0,0 +1,301 @@
+
+    pub fn subtract(&mut self, other: &ArrayRangeSet) {
+        // TODO: This can potentially be made more efficient, since the we know
+        // individual ranges are not overlapping, and the next range must start
+        // where the last one finished
+        for range in &other.0 {
+            self.remove(range.clone());
+        }
+    }
+
+    pub fn insert_one(&mut self, x: u64) -> bool {
+        self.insert(x..x + 1)
+    }
+
+    pub fn insert(&mut self, mut x: Range<u64>) -> bool {
+        let mut result = false;
+
+        let mut idx = 0;
+        while idx != self.0.len() {
+            let range = &mut self.0[idx];
+
+            if range.start > x.end {
+                // The range is fully before this range and therefore not extensible.
+                // Add a new range to the left
+                self.0.insert(idx, x);
+                return true;
+            } else if range.start > x.start {
+                // The new range starts before this range but overlaps.
+                // Extend the current range to the left
+                // Note that we don't have to merge a potential left range, since
+                // this case would have been captured by merging the right range
+                // in the previous loop iteration
+                result = true;
+                range.start = x.start;
+                // Modify the range to the start of this range, so that we can
+                // afterwards check for extending to the right
+                #[allow(clippy::almost_swapped)]
+                {
+                    x.start = range.start;
+                }
+                x.end = x.end.max(range.end);
+
+                if x.start >= x.end {
+                    return result;
+                }
+            }
+
+            // At this point we have handled all parts of the new range which
+            // are in front of the current range. Now we handle everything from
+            // the start of the current range
+
+            if x.start >= range.start && x.end <= range.end {
+                // Fully contained
+                return result;
+            } else if x.start <= range.end {
+                // Extend the current range to the end of the new range.
+                // Since it's not contained it must be bigger
+                range.end = x.end;
+
+                // Merge all follow-up ranges which are fully contained
+                while idx != self.0.len() - 1 {
+                    let curr = self.0[idx].clone();
+                    let next = self.0[idx + 1].clone();
+                    if curr.end >= next.start {
+                        self.0[idx].end = next.end.max(curr.end);
+                        self.0.remove(idx + 1);
+                    } else {
+                        break;
+                    }
+                }
+
+                return true;
+            }
+
+            idx += 1;
+        }
+
+        if x.start == x.end {
+            return result;
```

Comment:
Should this be `return false;`? The only case where we assign `result = true` is when the new range overlaps the beginning of an existing range, and in that case it looks like we always either judge the remainder wholly contained and return `true`, or extend the current range and return `true`, with control never reaching this point regardless.

Reply:
I guess that's true. But I don't think its worthwhile describing that this can only be false due to preconditions. It seems easier to just return the variable as is and knowing that it had been set correctly.

Reply:
Fair enough.